src/iq_stream_decoder.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2022 Transit Live Mapping Solutions
+ * All rights reserved.
+ *
+ * Authors:
+ *   Marenz Schmidl
+ *   Tassilo Tanneberger
+ */
+
+#include <thread>
+
+#include <iq_stream_decoder.hpp>
+
+IQStreamDecoder::IQStreamDecoder(std::shared_ptr<LowerMac> lower_mac,
+                                 std::shared_ptr<BitStreamDecoder> bit_stream_decoder, bool is_uplink)
+    : lower_mac_(lower_mac)
+    , bit_stream_decoder_(bit_stream_decoder)
+    , is_uplink_(is_uplink) {
+    std::transform(training_seq_n_.crbegin(), training_seq_n_.crend(),
+                   std::back_inserter(training_seq_n_reversed_conj_), [](auto v) { return std::conj(v); });
+    std::transform(training_seq_p_.crbegin(), training_seq_p_.crend(),
+                   std::back_inserter(training_seq_p_reversed_conj_), [](auto v) { return std::conj(v); });
+    std::transform(training_seq_x_.crbegin(), training_seq_x_.crend(),
+                   std::back_inserter(training_seq_x_reversed_conj_), [](auto v) { return std::conj(v); });
+
+    threadPool_ = std::make_shared<StreamingOrderedOutputThreadPoolExecutor<std::vector<std::function<void()>>>>(16);
+
+    upperMacWorkerThread_ = std::thread(&IQStreamDecoder::upperMacWorker, this);
+}
+
+void IQStreamDecoder::upperMacWorker() {
+    while (true) {
```

Comment:
maybe we need a way to gracefully kill this. So some variable that we can toggle if we want a graceful shutdown. Like when we receive some signal